pkg/kubelet/kubelet_server_journal.go

```diff
@@ -0,0 +1,395 @@
+
+	for _, service := range n.Services {
+		// Check the journalctl output to figure if the service is using journald or not. This is not needed in the
+		// Get-WinEvent case as the command returns an error if a service is not logging to the Application provider.
+		if checkForNativeLogger(ctx, service) {
+			nativeLoggers = append(nativeLoggers, service)
+		} else {
+			fileLoggers = append(fileLoggers, service)
+		}
+	}
+	return nativeLoggers, fileLoggers
+}
+
+// copyServiceLogs invokes journalctl or Get-WinEvent with the provided args. Note that
+// services are explicitly passed here to account for the heuristics.
+func (n *nodeLogQuery) copyServiceLogs(ctx context.Context, w io.Writer, services []string, previousBoot int) {
+	cmdStr, args, err := getLoggingCmd(n, services)
+	if err != nil {
+		fmt.Fprintf(w, "\nfailed to get logging cmd: %v\n", err)
+		return
+	}
+	cmd := exec.CommandContext(ctx, cmdStr, args...)
+	cmd.Stdout = w
+	cmd.Stderr = w
+
+	if err := cmd.Run(); err != nil {
+		if _, ok := err.(*exec.ExitError); ok {
+			return
+		}
+		if previousBoot == 0 {
+			fmt.Fprintf(w, "error: journal output not available\n")
+		}
+	}
+}
+
+// copyFileLogs loops over all the services and attempts to collect the file logs of each service
+func copyFileLogs(ctx context.Context, w io.Writer, services []string) {
+	if ctx.Err() != nil {
+		return
+	}
+
+	for _, service := range services {
+		heuristicsCopyFileLogs(ctx, w, service)
+	}
+}
+
+// heuristicsCopyFileLogs attempts to collect logs from either
+// /var/log/service
+// /var/log/service.log or
+// /var/log/service/service.log or
+// /var/log/service*INFO.* or
+// /var/log/service*INFO or
+// /var/log/service/service*INFO.* or
+// /var/log/service/service*INFO
+// in that order stopping on first success.
+func heuristicsCopyFileLogs(ctx context.Context, w io.Writer, service string) {
+	logDir := "/var/log/"
+	logFiles := [7]string{
+		filepath.Join(logDir, fmt.Sprintf("%s", service)),
+		filepath.Join(logDir, fmt.Sprintf("%s.log", service)),
+		filepath.Join(logDir, service, fmt.Sprintf("%s.log", service)),
+		filepath.Join(logDir, fmt.Sprintf("%s*INFO.*", service)),
+		filepath.Join(logDir, fmt.Sprintf("%s*INFO", service)),
+		filepath.Join(logDir, service, fmt.Sprintf("%s*INFO.*", service)),
+		filepath.Join(logDir, service, fmt.Sprintf("%s*INFO", service)),
+	}
+
+	var err error
+	for _, logFile := range logFiles {
+		if err = heuristicsCopyFileLog(w, logFile); err == nil {
+			break
+		}
+	}
+
+	if err != nil {
+		// If the last error was file not found it implies that no log files were for found for the service
+		if errors.Is(err, os.ErrNotExist) {
+			fmt.Fprintf(w, "\nlog not found for %s\n", service)
+			return
+		}
+		fmt.Fprintf(w, "\nerror getting logs for %s: %v\n", service, err)
+	}
+}
+
+// heuristicsCopyFileLog returns the contents of the given logFile
+func heuristicsCopyFileLog(w io.Writer, logFile string) error {
+	data, err := os.ReadFile(logFile)
```

Comment:
This function may need an update.

How large is this file? Using an io.Copy() would be more efficient.

Reply:
agreed, this line is an OOM waiting to happen

Reply:
Sorry folks, I am iterating really fast on this PR and meant to come back and fix this. Working on it now.

Reply:
I moved to using a context aware io.Reader().